[CSOptimizer] A few improvements to inference and ranking

xedin opened the pull request:
- Improve contextual result type handling during overload matching
- Infer result types through ternary expressions
- Infer argument candidates through optionals
- Allow matching against `CGFloat` as a contextual result type
- Expand initializer result type inference to all type and not just `Double` and `CGFloat`
- Add support for `??` operator
- Emulate old hack for unary argument matching more precisely
- Attempt choices marked as disfavored at the end of partition
- Expand literal support to bools, strings and dictionaries
- Support non-operator generic choices with simple signatures
- Introduce a way to prefer disjunctions before scores are considered (helps avoid selecting operators too aggressively)

<!--
If this pull request is targeting a release branch, please fill out the
following form:
https://github.com/swiftlang/.github/blob/main/PULL_REQUEST_TEMPLATE/release.md?plain=1

Otherwise, replace this comment with a description of your changes and
rationale. Provide links to external references/discussions if appropriate.
If this pull request resolves any GitHub issues, link them like so:

  Resolves <link to issue>, resolves <link to another issue>.

For more information about linking a pull request to an issue, see:
https://docs.github.com/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue
-->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[CSOptimizer] Improve contextual result type handling during overload matching

Result type should only be matched if there are matches on arguments
or there are no viable candidates.

xedin committed:
[CSOptimizer] Infer result types through ternary expressions

xedin committed:
[CSOptimizer] Infer argument candidates through optionals

For example, `??` operator could produce an optional type
so `test(<<something>> ?? 0) could result in an optional
argument that wraps a type variable. It should be possible
to infer bindings from underlying type variable and restore
optionality.

xedin committed:
[CSOptimizer] Allow matching against CGFloat as a contextual result type

xedin committed:
[CSOptimizer] Infer types from init calls used as arguments

This used to be limited to Double/CGFloat and operator arguments
but it's safe to do in general.

xedin committed:
[CSOptimizer] Don't reduce the ranking for non-default literal types

Since each candidate and overload choice are considered independenty
there is no way to judge whether non-default literal type is going
to result in a worse solution than non-default one.

xedin committed:
[CSOptimizer] Add support for `??` operator

xedin committed:
[Tests] NFC: Add a test-case with literal chain passed as an argument to a constructor

xedin committed:
[CSOptimizer] Emulate old hack for unary argument matching more precisely

Having it be part of the other matching wasn't a good idea because
previous "favoring" happened only in a few situations - if argument
was a declaration reference, application or (dynamic) subscript that
had overload choice selected during constraint generation.

xedin committed:
[CSOptimizer] Always prefer a disjunction with a single active chocie

Disjunction with a single element are not always introduced after
disfavoring, so we need to make sure that they are always preferred
during disjunction selection.

xedin committed:
[CSSolver] Attempt choices marked as disfavored at the end of partition

Since such choices are all but guaranteed to be worst than any
other non-disfavored choice, let's attempt them last to avoid
having to form a complete solution just to filter it out during
ranking.

xedin committed:
[CSOptimizer] Expand literal support to bools, strings and dictionaries

Optimizer now covers all of the most common ExpressibleBy*Literal
protocols.

xedin committed:
[CSOptimizer] Support non-operator generic choices with simple signatures

If there are no-same type requirements and parameters use
either concrete types or generic parameter types directly,
the optimizer should be able to handle ranking. Currently
candidate arguments are considered in isolation which makes
it impossible to deal with same-type requirements and
complex generic signatures.

xedin committed:
[CSOptimizer] Introduce a way to prefer disjunctions before scores are considered

Some of the disjunctions are not supported by the optimizers but
could still be a better choice than an operator. Using a non-score
based preference mechanism first allows us to make sure that
operator disjunctions are not selected too eagerly in some situations
when i.e. a member (supported or not) could be a better choice.

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

slavapestov commented on lib/Sema/CSOptimizer.cpp:
```diff
@@ -110,11 +195,47 @@ static bool isArithmeticOperator(ValueDecl *decl) {
+  // that use only concrete types or generic parameters directly
+  // in their parameter positions i.e. `(T, Int)`.
+
+  auto *paramList = getParameterList(decl);
```
Since you have a `GenericFunctionType` here, maybe just look at `choiceType->getParams()`?

xedin replied:
I think that would mean that I would have to look through self for instance members and I just didn’t want to mess with that…

slavapestov commented on lib/Sema/CSOptimizer.cpp:
```diff
@@ -281,6 +470,52 @@ static void findFavoredChoicesBasedOnArity(
+    return {/*score=*/0};
+
+  auto argumentType = cs.getType(argument);
+  if (argumentType->hasTypeVariable() || argumentType->hasDependentMember())
```
Is it possible for hasDependentMember() to return true while hasTypeVariable() returns false?

xedin replied:
Yeah, solver would form dependent member types when base is invalid as well

xedin committed:
[CSOptimizer] Disfavor operators some non-generic choices over all-generic standard comparison ones

Disfavor standard comparison operators when other wise has some
non-generic favored choices. Operators like `==` and `!=`
tend to get multiple generic matches through conformance to `Equatable`
and it's better to attempt them as a last resort.

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility
